streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -612,35 +598,22 @@ public void onAssignment(Assignment assignment) {
             assignedPartitions.add(partition);
         }
 
-        this.partitionsByHostState = info.partitionsByHost;
-
-        final Collection<Set<TopicPartition>> values = partitionsByHostState.values();
         final Map<TopicPartition, PartitionInfo> topicToPartitionInfo = new HashMap<>();
-        for (Set<TopicPartition> value : values) {
+        for (Set<TopicPartition> value : info.partitionsByHost.values()) {
             for (TopicPartition topicPartition : value) {
                 topicToPartitionInfo.put(topicPartition, new PartitionInfo(topicPartition.topic(),
-                                                                           topicPartition.partition(),
-                                                                           null,
-                                                                           new Node[0],
-                                                                           new Node[0]));
+                        topicPartition.partition(),
+                        null,
+                        new Node[0],
+                        new Node[0]));
             }
         }
-        metadataWithInternalTopics = Cluster.empty().withPartitions(topicToPartitionInfo);
 
-        checkForNewTopicAssignments(assignment);
-    }
+        taskManager.setClusterMetadata(Cluster.empty().withPartitions(topicToPartitionInfo));
```

Comment:
Could we encapsulate these method calls into a single method on `taskManager`? `updateAssignment` or something like that?

Reply:
Yeah I have also thought about this along the lines when trying to encapsulate TaskManager in an interface :) The problem is that except `setAssignmentMetadata` and `updateSubscriptionsFromAssignment` other two setters are also called in another place in StreamPartitionAssignor so we cannot really remove these functions as public APIs but rather replace `setAssignmentMetadata` with another function.

Thinking about it a bit more I think maybe it is still worth doing. Let me see what I can do here.